Index: modules/ssl/mod_ssl.c
===================================================================
--- modules/ssl/mod_ssl.c	(revision 1564310)
+++ modules/ssl/mod_ssl.c	(working copy)
@@ -37,6 +37,9 @@
 int ssl_running_on_valgrind = 0;
 #endif
 
+AP_IMPLEMENT_OPTIONAL_HOOK_RUN_ALL(int,ssl_new_client_pre,(server_rec *s,conn_rec *c,SSL *ssl),
+                                   (s,c,ssl),OK,DECLINED);
+
 /*
  *  the table of configuration directives we provide
  */
@@ -498,6 +501,8 @@
         return DECLINED; /* XXX */
     }
 
+    ap_run_ssl_new_client_pre(server, c, ssl);
+
     vhost_md5 = ap_md5_binary(c->pool, (unsigned char *)sc->vhost_id,
                               sc->vhost_id_len);
 
Index: modules/ssl/ssl_engine_init.c
===================================================================
--- modules/ssl/ssl_engine_init.c	(revision 1564310)
+++ modules/ssl/ssl_engine_init.c	(working copy)
@@ -29,6 +29,11 @@
 #include "ssl_private.h"
 #include "mpm_common.h"
 
+AP_IMPLEMENT_OPTIONAL_HOOK_RUN_ALL(int,ssl_server_init,(server_rec *s,SSL_CTX *ctx,apr_array_header_t *cert_files),
+                                   (s,ctx,cert_files),OK,DECLINED)
+AP_IMPLEMENT_OPTIONAL_HOOK_RUN_ALL(int,ssl_init_ctx,(server_rec *s,apr_pool_t *p,apr_pool_t *ptemp,int is_proxy,SSL_CTX *ctx),
+                                   (s,p,ptemp,is_proxy,ctx),OK,DECLINED);
+
 /*  _________________________________________________________________
 **
 **  Module Initialization
@@ -248,6 +253,26 @@
         return rv;
     }
 
+    for (s = base_server; s; s = s->next) {
+        sc = mySrvConfig(s);
+
+        if (sc->enabled == SSL_ENABLED_TRUE) {
+            if (sc->server && sc->server->ssl_ctx && sc->server->pks) {
+                if ((rv = ap_run_ssl_server_init(s, sc->server->ssl_ctx,
+                                                 sc->server->pks->cert_files)) != APR_SUCCESS) {
+                    ap_log_error(APLOG_MARK, APLOG_ERR, 0, s, APLOGNO()
+                                 "Init: ssl_server_init failed");
+                    return rv;
+                }
+            }
+            else {
+                ap_log_error(APLOG_MARK, APLOG_CRIT, 0, s, APLOGNO()
+                             "Init: some interesting stuff is missing");
+                return HTTP_INTERNAL_SERVER_ERROR;
+            }
+        }
+    }
+
     /*
      *  Announce mod_ssl and SSL library in HTTP Server field
      *  as ``mod_ssl/X.X.X OpenSSL/X.X.X''
@@ -815,6 +840,12 @@
 #endif
     }
 
+    if ((rv = ap_run_ssl_init_ctx(s, p, ptemp, 
+                                  mctx->sc && mctx->sc->proxy_enabled,
+                                  mctx->ssl_ctx)) != APR_SUCCESS) {
+        return rv;
+    }
+
     return APR_SUCCESS;
 }
 
Index: modules/ssl/ssl_engine_io.c
===================================================================
--- modules/ssl/ssl_engine_io.c	(revision 1564310)
+++ modules/ssl/ssl_engine_io.c	(working copy)
@@ -31,6 +31,9 @@
 #include "mod_ssl.h"
 #include "apr_date.h"
 
+AP_IMPLEMENT_OPTIONAL_HOOK_RUN_ALL(int,ssl_new_client,(server_rec *s,conn_rec *c),
+                                   (s,c),OK,DECLINED)
+
 /*  _________________________________________________________________
 **
 **  I/O Hooks
@@ -1104,6 +1107,7 @@
     X509 *cert;
     int n;
     int ssl_err;
+    int rc;
     long verify_result;
     server_rec *server;
 
@@ -1350,6 +1354,12 @@
         return APR_ECONNABORTED;
     }
 
+    if ((rc = ap_run_ssl_new_client(server, c)) != OK && rc != DECLINED) {
+        ap_log_cerror(APLOG_MARK, APLOG_ERR, rc, c,
+                      "ssl_new_client failed");
+        return rc;
+    }
+
     return APR_SUCCESS;
 }
 
Index: modules/ssl/ssl_hooks.h
===================================================================
--- modules/ssl/ssl_hooks.h	(revision 0)
+++ modules/ssl/ssl_hooks.h	(working copy)
@@ -0,0 +1,33 @@
+/* Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef SSL_HOOKS_H
+#define SSL_HOOKS_H
+
+/* OpenSSL headers */
+#include <openssl/opensslv.h>
+#if (OPENSSL_VERSION_NUMBER >= 0x10001000)
+/* must be defined before including ssl.h */
+#define OPENSSL_NO_SSL_INTERN
+#endif
+#include <openssl/ssl.h>
+
+AP_DECLARE_HOOK(int,ssl_server_init,(server_rec *,SSL_CTX *,apr_array_header_t *))
+AP_DECLARE_HOOK(int,ssl_new_client,(server_rec *,conn_rec *c))
+AP_DECLARE_HOOK(int,ssl_new_client_pre,(server_rec *,conn_rec *,SSL *ssl))
+AP_DECLARE_HOOK(int,ssl_init_ctx,(server_rec *s,apr_pool_t *p,apr_pool_t *ptemp,int is_proxy,SSL_CTX *ctx))
+
+#endif /* SSL_HOOKS_H */
Index: modules/ssl/ssl_private.h
===================================================================
--- modules/ssl/ssl_private.h	(revision 1564310)
+++ modules/ssl/ssl_private.h	(working copy)
@@ -932,6 +932,8 @@
 extern int ssl_running_on_valgrind;
 #endif
 
+#include "ssl_hooks.h"
+
 #endif /* SSL_PRIVATE_H */
 /** @} */
 
